src/target/riscv/riscv-013.c

```diff
@@ -3214,7 +3225,15 @@ static enum riscv_halt_reason riscv013_halt_reason(struct target *target)
 
 int riscv013_write_debug_buffer(struct target *target, unsigned index, riscv_insn_t data)
 {
-	return dmi_write(target, DMI_PROGBUF0 + index, data);
+	dm013_info_t *dm = get_dm(target);
+	if (dm->progbuf_cache[index] != data) {
+		if (dmi_write(target, DMI_PROGBUF0 + index, data) != ERROR_OK)
+			return ERROR_FAIL;
+		dm->progbuf_cache[index] = data;
+	} else {
+		LOG_DEBUG("cache hit for 0x%x @%d", data, index);
```

Comment:
So during a load sequence, there is enough commonality between successive progbuf contents that caching is a noticeable performance difference?  Makes sense.

Reply:
I wouldn't call it noticeable, but it is measurable.

Flashing already uses an on-chip algorithm to speed things up, so there's a bunch of overhead in reading/writing registers and halting/resuming to make that go for every 4K(?) block. The second word of the program buffer is very often set to ebreak, and I think that's where most of the cache hits come in. I didn't really look, though.